File: Problems/Algorithms/1971. Find If Path Exists in Graph/find_path.py
#!/usr/bin/env python3

from collections import defaultdict
from typing import List


class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = defaultdict(set)
        for edge in edges:
            x, y = edge
            graph[x].add(y)
            graph[y].add(x)

        visited = set()
        queue = [source]
        while len(queue) > 0:
            curr = queue.pop(0)
            if curr not in visited:
                if curr == destination:
                     return True
                for neighbor in graph[curr]:
                    if neighbor not in visited:
                        queue.append(neighbor)
                visited.add(curr)

        return False

File: Problems/Algorithms/1971. Find If Path Exists in Graph/find-path.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>

using namespace std;

class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        unordered_map<int, vector<int>> graph;
        for (auto& edge : edges) {
            int u = edge[0], v = edge[1];
            graph[u].push_back(v);
            graph[v].push_back(u);
        }

        unordered_set<int> visited;
        visited.insert(source);

        queue<int> queue;
        queue.push(source);
        while (!queue.empty()) {
            int node = queue.front();
            queue.pop();

            if (node == destination) {
                return true;
            }

            for (int neighbor : graph[node]) {
                if (visited.find(neighbor) == visited.end()) {
                    visited.insert(neighbor);
                    queue.push(neighbor);
                }
            }
        }

        return false;
    }
};
